### PR TITLE
`General`: Select entire line when using ordered/unordered lists in the Markdown editor

### PR DESCRIPTION
**Pull request by Dominik-Weinzierl**
#### General

- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Follow up of #5072: The selection when using ordered/unordered lists is extended to the entire line.

### Description

In order to increase the usability when using order/unordered lists, the selection is extended to the entire line.
In addition, this PR fixes some remaining issues with the unordered list.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Exercise

1. Log in to Artemis
2. Navigate to a Markdown editor (e.g. exercise details)
3. Insert multiple lines and create an ordered/unordered list (select only half of a line)
4. Select the ordered/unordered list and revert it back to text
5. Insert empty lines and create again an ordered/unordered list
6. Check that everything works as expected (play around)

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| orderedListCommand.ts | 100% | 100% |
| unorderedListCommand.ts | 100% | 100% |

### Screenshots

https://user-images.githubusercontent.com/60005702/168684075-0d9a3cc6-46b5-479c-910d-c60b7a150c5e.mp4

### CODE REVIEW

**Review by xlieA — APPROVED** [review]
I tested on ts3 and everything worked as described.

**Review by akesfeden — APPROVED** [review]
Tested on TS1, works as described, thanks for following up on this topic 👍 Code changes also look good.

**Review by TheTrueVester — APPROVED** [review]
Tested on ts1. Everything works as expected. If you however swap between clicking unordered and ordered list, you could end up with a mess that keeps on growing. This however doesn't really need to be fixed.
image

File: src/main/webapp/app/shared/markdown-editor/commands/command.ts
Comment (bassner) [inline src/main/webapp/app/shared/markdown-editor/commands/command.ts:5]: Can we use  ``import`` instead? We're not using ``require`` anywhere else in our application code

**Review by bassner — DISMISSED** [review]
Just one comment / question

File: src/main/webapp/app/shared/markdown-editor/commands/command.ts
Comment (Dominik-Weinzierl) [reply src/main/webapp/app/shared/markdown-editor/commands/command.ts:5]: I guess that should also work :D

**Review by mallgaier — APPROVED** [review]
Tested on TS3, works as expected!
Good job 👍

**Review by TheTrueVester — DISMISSED** [review]
Tested on ts2, everything worked as expected

File: src/main/webapp/app/shared/markdown-editor/commands/command.ts
Comment (b-fein) [inline src/main/webapp/app/shared/markdown-editor/commands/command.ts:42]:
```suggestion
    /**
     * Extends the current selection to full lines.
     *
     * @return The complete lines of the selected text.
     */
    protected getExtendedSelectedText(): string[] {
```

File: src/main/webapp/app/shared/markdown-editor/commands/orderedListCommand.ts
Comment (b-fein) [inline src/main/webapp/app/shared/markdown-editor/commands/orderedListCommand.ts:20]:
```suggestion
     *
```
I don’t think this is needed as part of the JSdoc any more.

File: src/main/webapp/app/shared/markdown-editor/commands/orderedListCommand.ts
Comment (b-fein) [inline src/main/webapp/app/shared/markdown-editor/commands/orderedListCommand.ts:23]: This method should be renamed, as it no longer splits the text, but only triggers the manipulation for each line.

The same applies to the method with the same name in the unordered list.

**Review by b-fein — COMMENTED** [review]
I’ve only got small suggestions for the code.

I did not test the functionality (yet).

**Review by nadinesamyy — DISMISSED** [review]
Tested on ts1, works as described

**Review by b-fein — DISMISSED** [review]
Thanks, works great. :+1: 

Code looks good, too.

**Review by JohannesStoehr — COMMENTED** [review]
Tested on TS1 and found these two things.
If they are not relevant or addressed in a Follow up, see this comment as a approve.

When selecting an ordered line and a line just with text and clicking on "ordered list" the second line starts with a 2., even though there is no longer a 1. above:
https://user-images.githubusercontent.com/38322605/169580591-3c3e7aaa-d3cb-4199-a5cd-b7cec7d20c94.mov

When selecting multiple line but leaving out some in between the list buttons only affect one of the selections:
https://user-images.githubusercontent.com/38322605/169580570-9020e704-45cb-4520-ac29-cd9d44fc9d3b.mov

**Review by bottbenj — CHANGES_REQUESTED** [review]
There is no need to define the range classes, brace does have types available. (also even if you defined them here instead of using the existing ones, they should be interfaces instead of classes)
```diff
diff --git a/src/main/webapp/app/shared/markdown-editor/commands/command.ts b/src/main/webapp/app/shared/markdown-editor/commands/command.ts 
index 18192ad21d..631afc7c45 100644
--- a/src/main/webapp/app/shared/markdown-editor/commands/command.ts
+++ b/src/main/webapp/app/shared/markdown-editor/commands/command.ts
@@ -1,20 +1,9 @@
 import { ElementRef } from '@angular/core';
 import { IconProp } from '@fortawesome/fontawesome-svg-core';
-import ace from 'brace';
+import { acequire, Range } from 'brace';

-// Work around to update the range
-const Range = ace.acequire('ace/range').Range;
-
-// Required to access end and start
-class RangeContainer {
-    end: Position;
-    start: Position;
-}
-
-class Position {
-    row: number;
-    column: number;
-}
+// The imported range isn't actually a constructor so get it like this
+const RangeCtor = acequire('ace/range').Range as typeof Range;

 /**
  * abstract class for all commands - default and domain commands of Artemis
@@ -50,10 +39,10 @@ export abstract class Command {
         // Split text by line breaks
         const lines = text.split('\n');

-        const range = this.getRange() as unknown as RangeContainer;
+        const range = this.getRange();

         // Update the range
-        this.aceEditor.selection.setRange(new Range(range.start.row, 0, range.end.row, lines[range.end.row].length));
+        this.aceEditor.selection.setRange(new RangeCtor(range.start.row, 0, range.end.row, lines[range.end.row].length));

         // Return extended selection as array
         return lines.slice(range.start.row, range.end.row + 1);
```

If there is a good reason that the types are not used and the `aceEditor` is declared as `any` instead of `Editor` then the existing `Range` annotations should be removed instead as they (as you noticed) refer to the wrong Range type. If not the `Editor` type should be used, but that happens in a whole bunch of places so i'd leave that to its own PR.

**Comment by Dominik-Weinzierl** [comment]
@JohannesStoehr yes the markdown editor and especially the list functionality has many edge cases. We can evaluate in another PR which one we want to handle and which one we want to "ignore" :thinking:

**Review by bottbenj — APPROVED** [review]
Works as expected, Code looks good.

The tests could be simplified/deduplicated by using parameterized tests, but I don't think that is required.

**Review by b-fein — APPROVED** [review]
Re-approve after code changes.

**Review by rhubinger — APPROVED** [review]
Tested on TS2. Works as expected.

**Review by JohannesStoehr — APPROVED** [review]
All the core functionality works 👍

**Pull request merged**
